pyserini/search/__main__.py

```diff
@@ -131,36 +133,52 @@
+        else:
+            batch_topics.append(topic)
+            if (index + 1) % args.batch_size == 0 or index == len(topics.keys()) - 1:
+                queries = [topics[topic].get('title') for topic in batch_topics]
+                ids = [str(topic) for topic in batch_topics]
+                results = searcher.batch_search(queries, ids, args.hits, args.threads)
+                results = sorted(results.items(), key=lambda item: int(item[0]))
+                batch_topics.clear()
+            else:
+                continue
 
-            if use_prcl and len(hits) > (args.r + args.n):
-                scores, docids = ranker.rerank(docids, scores)
+        for result in results:
+            topic, hits = result
+            if not args.max_passage:
+                docids = [hit.docid.strip() for hit in hits]
+                scores = [hit.score for hit in hits]
 
-            if args.msmarco:
-                for i, docid in enumerate(docids):
-                    target_file.write(f'{topic}\t{docid}\t{i + 1}\n')
-            else:
-                for i, (docid, score) in enumerate(zip(docids, scores)):
-                    target_file.write(f'{topic} Q0 {docid} {i + 1} {score:.6f} {tag}\n')
-        else:
-            unique_docs = set()
-            rank = 1
-            for hit in hits:
-                docid, _ = hit.docid.split(args.max_passage_delimiter)
-                if docid in unique_docs:
-                    continue
+                if use_prcl and len(hits) > (args.r + args.n):
+                    scores, docids = ranker.rerank(docids, scores)
 
                 if args.msmarco:
-                    target_file.write(f'{topic}\t{docid}\t{rank}\n')
+                    for i, docid in enumerate(docids):
+                        target_file.write(f'{topic}\t{docid}\t{i + 1}\n')
                 else:
-                    target_file.write(f'{topic} Q0 {docid} {rank} {hit.score:.6f} {tag}\n')
-                rank = rank + 1
-                unique_docs.add(docid)
-                if rank > args.max_passage_hits:
-                    break
+                    for i, (docid, score) in enumerate(zip(docids, scores)):
+                        target_file.write(f'{topic} Q0 {docid} {i + 1} {score:.6f} {tag}\n')
+            else:
+                unique_docs = set()
+                rank = 1
+                for hit in hits:
+                    docid, _ = hit.docid.split(args.max_passage_delimiter)
+                    if docid in unique_docs:
+                        continue
+
+                    if args.msmarco:
+                        target_file.write(f'{topic}\t{docid}\t{rank}\n')
+                    else:
+                        target_file.write(f'{topic} Q0 {docid} {rank} {hit.score:.6f} {tag}\n')
+                    rank = rank + 1
+                    unique_docs.add(docid)
+                    if rank > args.max_passage_hits:
+                        break
+        results.clear()
```

Comment:
Should I refactor this into a seperate function? Which takes a "result", Tuple[str, List[JSimpleSearcherResult]], and writes it to the output file.

Reply:
I think the most important top-level separate is whether we're doing max-passage or not?

Reply:
So best refactor it as two functions which write a "result", `write_result_max_passage` and `write_result`?

Reply:
I think so... give it a try and let's see how it looks?